test/Microsoft.AspNetCore.Razor.Test.Common/Language/IntegrationTests/IntegrationTestBase.cs

```diff
@@ -220,9 +220,102 @@ protected void AssertSourceMappingsMatchBaseline(RazorCodeDocument document)
+                    continue;
+                }
+
+                var location = new SourceSpan(span.Start, span.Length);
+                var expectedSpan = sourceContent.Substring(span.Start.AbsoluteIndex, span.Length);
+
+                // See #
+                if (string.IsNullOrWhiteSpace(expectedSpan))
+                {
+                    // For now we don't verify whitespace inside of a directive. We know that directives cheat
+                    // with how they bound whitespace/C#/markup to make completion work.
+                    if (span.Parent is Block block && block.Type == BlockKindInternal.Directive)
+                    {
+                        continue;
+                    }
+                }
+
+                // See #
+                if (string.Equals("@", expectedSpan) && span.Kind == SpanKindInternal.Code)
+                {
+                    // For now we don't verify an escaped transition. In some cases one of the @ tokens in @@foo
+                    // will be mapped as C# but will not be present in the output buffer because it's not actually C#.
+                    continue;
+                }
+
+                var found = false;
+                for (var j = 0; j < csharpDocument.SourceMappings.Count; j++)
+                {
+                    var mapping = csharpDocument.SourceMappings[j];
+                    if (mapping.OriginalSpan == location)
+                    {
+                        var actualSpan = csharpDocument.GeneratedCode.Substring(
+                            mapping.GeneratedSpan.AbsoluteIndex, 
+                            mapping.GeneratedSpan.Length);
+
+                        if (!string.Equals(expectedSpan, actualSpan, StringComparison.Ordinal))
+                        {
+                            throw new XunitException(
+                                $"Found the span {location} in the output mappings but it contains " +
+                                $"'{EscapeWhitespace(actualSpan)}' instead of '{EscapeWhitespace(expectedSpan)}'.");
+                        }
+
+                        found = true;
+                        break;
+                    }
+                }
+
+                if (!found)
+                {
+                    throw new XunitException(
+                        $"Could not find the span {location} - containing '{EscapeWhitespace(expectedSpan)}' " +
+                        $"in the output.");
+                }
+            }
+        }
+
+        private class CodeSpanVisitor : ParserVisitor
+        {
+            public List<Span> CodeSpans { get; } = new List<Span>();
+
+            public override void VisitSpan(Span span)
+            {
+                if (span.Kind == SpanKindInternal.Code)
+                {
+                    CodeSpans.Add(span);
+                }
+
+                base.VisitSpan(span);
+            }
+        }
+
+        private static string EscapeWhitespace(string content)
```

Comment:
nest this function in the verify method so we don't accidentally use this in the future. Our other tests should properly already escape whitespace.

Reply:
That wouldn't be harmful though?